lib/translator/command_translator.rb

```diff
@@ -409,16 +469,40 @@ def pop_kernel_launcher(command_translation_result)
             # a new kernel will be built.
             def translate_input(input)
                 if input.pattern == :tid
-                    # Stay in current kernel
-                    return input.command.accept(self)
+                    # Stay in current kernel                    
+                    if input.command.gpu_result_pointer.nil? || input.command.gpu_result_pointer.device_pointer == 0
+                        result = input.command.accept(self)
+                    else
```

Comment:
Can we move that nested `if` outside of the outer `if`? Maybe then we don't have to duplicate it down there?